### PR TITLE
Implement account.viewState which allows querying contract state directly

### PR DESCRIPTION
**Pull request by vgrichina**

### CODE REVIEW

File: src/account.ts [hunk 1]
```diff
@@ -373,6 +374,19 @@ export class Account {
         return result.result && result.result.length > 0 && parse(Buffer.from(result.result));
     }
 
+    async viewState(prefix: string | Uint8Array, blockQuery: { blockId: BlockId } | { finality: Finality } ) {
+        const { blockId, finality } = blockQuery as any || {};
+        const { values } = await this.connection.provider.query({
+            request_type: 'view_state', 
+            block_id: blockId,
+            finality: blockId ? undefined : finality || 'optimistic',
```
Comment (chadoh) [inline src/account.ts:382]: There's a pretty tricky interplay between `blockId` & `finality` here that I'd like to see documented more explicitly. Can we add a code comment above `viewState` to get picked up by typedoc, as well as throwing an error message when you pass a nonsensical combination of the two?

File: src/account.ts [hunk 2]
```diff
@@ -373,6 +374,19 @@ export class Account {
...
+            account_id: this.accountId,
+            prefix_base64: Buffer.from(prefix,).toString('base64')
```
Comment (chadoh) [inline src/account.ts:384]: What's with the comma in `(prefix,)`?

File: src/account.ts [hunk 3]
```diff
@@ -373,6 +374,19 @@ export class Account {
...
+        });
+
+        return values.map(({key, value}) => ({ key: Buffer.from(key, 'base64'), value: Buffer.from(value, 'base64')}))
```
Comment (chadoh) [inline src/account.ts:387]: Please break up this long line

File: src/providers/json-rpc-provider.ts [hunk 4]
```diff
@@ -68,13 +68,17 @@ export class JsonRpcProvider extends Provider {
 
     /**
      * Query the RPC as [shown in the docs](https://docs.nearprotocol.com/docs/interaction/rpc#query)
-     * @param path Path parameter for the RPC (ex. "contract/my_token")
-     * @param data Data parameter (ex. "", "AQ4", or whatever is needed)
```
Comment (chadoh) [inline src/providers/json-rpc-provider.ts:72]: Please add documentation to describe the new interface & behavior

File: src/providers/provider.ts [hunk 5]
```diff
@@ -241,6 +242,7 @@ export abstract class Provider {
 
     abstract async sendTransaction(signedTransaction: SignedTransaction): Promise<FinalExecutionOutcome>;
     abstract async txStatus(txHash: Uint8Array, accountId: string): Promise<FinalExecutionOutcome>;
+    abstract async query(params: object): Promise<any>;
     abstract async query(path: string, data: string): Promise<any>;
```
Comment (chadoh) [inline src/providers/provider.ts:246]: Do we want to have both versions of `query` here?

File: src/providers/provider.ts [hunk 6]
```diff
@@ -26,6 +26,7 @@ type BlockHash = string;
 type BlockHeight = number;
 export type BlockId = BlockHash | BlockHeight;
 
+// TODO: Remove near-final?
```
Comment (chadoh) [inline src/providers/provider.ts:29]: `near-final` used to be in the docs, but I don't see it anymore. Maybe this was removed. Who can we ask?

Seems like we could make a separate PR that just removes this option, and request review from people who would know enough to approve or reject the change.

**Review by chadoh — CHANGES_REQUESTED** [review]
Thanks for this! Excited to see it get in. I'd like to see some changes related to documentation & devx before it gets merged, but nothing that should take long. See below.

File: src/providers/json-rpc-provider.ts (hunk 4 above)
Comment (vgrichina) [reply src/providers/json-rpc-provider.ts:72]: @chadoh what inft do you think is going to be useful outside of link to corresponding RPC docs?

If I added a doc in style of the docs we had before it would be something like:
```
@param params parameters for the RPC
```

which I don't think is adding value

File: src/providers/provider.ts (hunk 5 above)
Comment (vgrichina) [reply src/providers/provider.ts:246]: Yes to avoid breaking backwards compatibility

File: src/providers/provider.ts (hunk 6 above)
Comment (vgrichina) [reply src/providers/provider.ts:29]: No idea, that's why I just added TODO

File: src/account.ts (hunk 1 above)
Comment (vgrichina) [reply src/account.ts:382]: yeah, I forgot about the doc here, need to add

Comment (vgrichina) [reply src/account.ts:382]: > as well as throwing an error message when you pass a nonsensical combination of the two?

Types kinda describe it. We aren't really checking types during runtime in most other places. TBH this is something I really don't understand about TypeScript. Like doing this whole effort with types and then not having easy way to check in runtime automatically 🤷‍♂️

Comment (vgrichina) [reply src/account.ts:382]: created issue to track https://github.com/near/near-api-js/issues/457

File: src/providers/provider.ts (hunk 6 above)
Comment (vgrichina) [reply src/providers/provider.ts:29]: Seems like this is secret feature 🤷‍♂️

**Comment by vgrichina** [comment]
Added docs and tests, etc @chadoh 
Going to merge
